File: contrib/ifunc/stop.c
/*
 * Stops the ifunc polling loop. TODO: pure?
 * Author: Luis E. P.
 * Copyright (C) 2021-2022 Arm Ltd. All rights reserved.
 *
 */

#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>

#include <ucp/api/ucp.h>

#include "pointer_chase.h"

size_t stop_payload_bound(void *source_args, size_t source_args_size)
{
	(void)source_args;
	return source_args_size;
}


int stop_payload_init(void *source_args,
						size_t source_args_size,
						void *payload,
						size_t *payload_size)
{
    (void)source_args;
    (void)source_args_size;
    (void)payload;
    (void)payload_size;
    return 0;
}


void stop_main(void *payload, size_t payload_size, void *target_args)
{
	(void)payload_size;
	(void)payload;

	pc_tgt_args_t *tgt_args = (pc_tgt_args_t *)target_args;

	tgt_args->run = false;
}

File: contrib/ifunc/julia/stop.jl
##
# Dynamic adaptive pointer chasing application
#
# Author: Valentin Churavy
# Massachussetts Institute of Technology, 2022
#
# Original implementation:
# Author: Luis E. P.
# Copyright (C) 2021-2022 Arm Ltd. All rights reserved.
#
##
using UCX

include("UCXAsap.jl")
import .UCXAsap

include("IFunc.jl")
import .IFunc 

include("unsafe.jl")
include("common.jl")

function bound(source_args::Ptr{Cvoid}, source_args_size::Csize_t)::Csize_t
	return source_args_size
end

function init(source_args::Ptr{Cvoid}, source_args_size::Csize_t, payload::Ptr{Cvoid}, payload_size::Csize_t)::Cint
    return Cint(0)
end

function main(payload::Ptr{Cvoid}, payload_size::Csize_t, target_args::Ptr{Cvoid})::Cvoid
    tgt_args = Base.unsafe_convert(Ptr{TgtArgs}, target_args)
    unsafe_setfield!(tgt_args, :run, false)
    return nothing
end

const fns = (
    (bound, Tuple{Ptr{Cvoid}, Csize_t}),
    (init, Tuple{Ptr{Cvoid}, Csize_t, Ptr{Cvoid}, Csize_t}),
    (main, Tuple{Ptr{Cvoid}, Csize_t, Ptr{Cvoid}})
)


IFunc.Compiler.emit(fns, "jstop", "$(Sys.ARCH)-unknown-linux-gnu")
